Add Lock View documents (#5798)
pingcap/docs: troubleshoot-lock-conflicts.md

Solutions:

* If it is difficult to confirm the cause of the deadlock, for v5.1 and later versions, you are recommended to try to query the `INFORMATION_SCHEMA.DEADLOCKS` or `INFORMATION_SCHEMA.CLUSTER_DEADLOCKS` system table to get the information of deadlock waiting chain. For details, see the [Deadlock error](#deadlock-error) section and the [`DEADLOCKS` table](/information-schema/information-schema-deadlocks.md) document.

> * If the deadlock occurs frequently, you need to adjust your application code to reduce such occurrences.	* If the deadlock occurs frequently, you need to adjust your application code to reduce such occurrences.

Author: ```suggestion
* If the deadlock occurs frequently, you need to adjust your application code to reduce such occurrences.
```

* [`DATA_LOCK_WAITS`](/information-schema/information-schema-data-lock-waits.md): Provides the pessimistic lock-waiting information in TiKV, including the `start_ts` of the blocking and blocked transaction, the digest of the blocked SQL statement, and the key on which the waiting occurs.
* [`DEADLOCKS` and `CLUSTER_DEADLOCKS`](/information-schema/information-schema-deadlocks.md): Provides the information of several deadlock events that have recently occurred on the current TiDB node or in the entire cluster, including the waiting relationship among transactions in the deadlock loops, the digest of the statement currently being executed in the transaction, and the key on which the waiting occurs.

> **Warning:**
>
> Currently, this is an experimental feature. The definition and behavior of the table structure might have major changes in future releases.

> The following sections show the examples of troubleshooting some issues using these tables.

Reviewer: ```suggestion

The following sections show the examples of troubleshooting some issues using these tables.
```

Without the empty line, it's included in the quote section, but it's supposed not.

{{< copyable "sql" >}}

```sql
select l.deadlock_id, l.occur_time, l.try_lock_trx_id, l.trx_holding_lock, s.digest_text from information_schema.deadlocks as l left join information_schema.statements_summary as s on l.current_sql_digest = s.digest;
```

```sql
+-------------+----------------------------+--------------------+--------------------+-----------------------------------------+
| deadlock_id | occur_time                 | try_lock_trx_id    | trx_holding_lock   | digest_text                             |
+-------------+----------------------------+--------------------+--------------------+-----------------------------------------+
|           1 | 2021-06-04 08:22:38.765699 | 425405959304904707 | 425405959304904708 | update `t` set `v` = ? where `id` = ? ; |
|           1 | 2021-06-04 08:22:38.765699 | 425405959304904708 | 425405959304904707 | update `t` set `v` = ? where `id` = ? ; |
+-------------+----------------------------+--------------------+--------------------+-----------------------------------------+
```

#### A few hot keys cause queueing locks

> The `DATA_LOCK_WAITS` system table provides the lock-waiting status on the TiKV node. When you query this table, TiDB automatically obtains the real-time lock-waiting information from all TiKV nodes. If a few hot keys are frequently locked and block many transactions, you can query the `DATA_LOCK_WAITS` table and aggregate the results by key to try to find the key on which issues frequently occur:

Reviewer: ```suggestion
The `DATA_LOCK_WAITS` system table provides the lock-waiting status on the TiKV nodes. When you query this table, TiDB automatically obtains the real-time lock-waiting information from all TiKV nodes. If a few hot keys are frequently locked and block many transactions, you can query the `DATA_LOCK_WAITS` table and aggregate the results by key to try to find the key on which issues frequently occur:
```

{{< copyable "sql" >}}

```sql
select l.key, trx.* from information_schema.data_lock_waits as l join information_schema.tidb_trx as trx on l.current_holding_trx_id = trx.id where l.trx_id = 425497223886536705\G
```

```sql
*************************** 1. row ***************************
               key: 7480000000000000475f728000000000000002
                ID: 425497219115778059
        START_TIME: 2021-06-08 09:04:38.292000
CURRENT_SQL_DIGEST: a4e28cc182bdd18288e2a34180499b9404cd0ba07e3cc34b6b3be7b7c2de7fe9
             STATE: LockWaiting
WAITING_START_TIME: 2021-06-08 09:04:38.336264
   MEM_BUFFER_KEYS: 1
  MEM_BUFFER_BYTES: 19
        SESSION_ID: 97
              USER: root
                DB: test
   ALL_SQL_DIGESTS: [0fdc781f19da1c6078c9de7eadef8a307889c001e05f107847bee4cfc8f3cdf3, a4e28cc182bdd18288e2a34180499b9404cd0
ba07e3cc34b6b3be7b7c2de7fe9, a4e28cc182bdd18288e2a34180499b9404cd0ba07e3cc34b6b3be7b7c2de7fe9]
1 row in set (0.01 sec)
```

> If the `start_ts` of the current transaction is unknown, you can try to find out the information of the blocking transaction in the `TIDB_TRX` / `CLUSTER_TIDB_TRX` table or [`PROCESSLIST` / `CLUSTER_PROCESSLIST`](/information-schema/information-schema-processlist.md) table.

Reviewer: ```suggestion
If the `start_ts` of the current transaction is unknown, you can try to find it out from the information in the `TIDB_TRX` / `CLUSTER_TIDB_TRX` table or [`PROCESSLIST` / `CLUSTER_PROCESSLIST`](/information-schema/information-schema-processlist.md) table.
```

Actually I mean trying to find out the `start_ts`  of the current transaction, and the current transaction is the blocked transaction rather than blocking transaction.

+-------------+----------------------------+-----------+--------------------+------------------------------------------------------------------+----------------------------------------+--------------------+
|           1 | 2021-06-04 08:22:38.765699 |         0 | 425405959304904707 | 22230766411edb40f27a68dadefc63c6c6970d5827f1e5e22fc97be2c4d8350d | 7480000000000000385F728000000000000002 | 425405959304904708 |
|           1 | 2021-06-04 08:22:38.765699 |         0 | 425405959304904708 | 22230766411edb40f27a68dadefc63c6c6970d5827f1e5e22fc97be2c4d8350d | 7480000000000000385F728000000000000001 | 425405959304904707 |
+-------------+----------------------------+-----------+--------------------+------------------------------------------------------------------+----------------------------------------+--------------------+
```

The query result above shows the waiting relationship among multiple transactions in the deadlock error, the digest of the SQL statement currently being executed in each transaction, and the key on which the conflict occurs.

> You can get the text of the normalized SQL statement corresponding to the digest of the SQL statements executed recently from the `STATEMENTS_SUMMARY` or `STATEMENTS_SUMMARY_HISTORY` table. For details, see [`STATEMENTS_SUMMARY` and `STATEMENTS_SUMMARY_HISTORY` tables](/statement-summary-tables.md). You can also join the obtained results directly with the `DEADLOCKS` table. Note: `STATEMENTS_SUMMARY` might not contain the information of all SQL statements, so left join is used in the following example:

Reviewer 2: ```suggestion
You can get the text of the normalized SQL statement corresponding to the digest of the SQL statements executed recently from the `STATEMENTS_SUMMARY` or `STATEMENTS_SUMMARY_HISTORY` table. For details, see [`STATEMENTS_SUMMARY` and `STATEMENTS_SUMMARY_HISTORY` tables](/statement-summary-tables.md). You can also join the obtained results directly with the `DEADLOCKS` table. Note that the `STATEMENTS_SUMMARY` table might not contain the information of all SQL statements, so left join is used in the following example:
```

{{< copyable "sql" >}}

```sql
select l.deadlock_id, l.occur_time, l.try_lock_trx_id, l.trx_holding_lock, s.digest_text from information_schema.deadlocks as l left join information_schema.statements_summary as s on l.current_sql_digest = s.digest;
```

```sql
+-------------+----------------------------+--------------------+--------------------+-----------------------------------------+
| deadlock_id | occur_time                 | try_lock_trx_id    | trx_holding_lock   | digest_text                             |
+-------------+----------------------------+--------------------+--------------------+-----------------------------------------+
|           1 | 2021-06-04 08:22:38.765699 | 425405959304904707 | 425405959304904708 | update `t` set `v` = ? where `id` = ? ; |
|           1 | 2021-06-04 08:22:38.765699 | 425405959304904708 | 425405959304904707 | update `t` set `v` = ? where `id` = ? ; |
+-------------+----------------------------+--------------------+--------------------+-----------------------------------------+
```

#### A few hot keys cause queueing locks

> The `DATA_LOCK_WAITS` system table provides the lock-waiting status on the TiKV nodes. When you query this table, TiDB automatically obtains the real-time lock-waiting information from all TiKV nodes. If a few hot keys are frequently locked and block many transactions, you can query the `DATA_LOCK_WAITS` table and aggregate the results by key to try to find the key on which issues frequently occur:

Reviewer 2: ```suggestion
The `DATA_LOCK_WAITS` system table provides the lock-waiting status on the TiKV nodes. When you query this table, TiDB automatically obtains the real-time lock-waiting information from all TiKV nodes. If a few hot keys are frequently locked and block many transactions, you can query the `DATA_LOCK_WAITS` table and aggregate the results by key to try to find the keys on which issues frequently occur:
```

{{< copyable "sql" >}}

```sql
select l.key, trx.* from information_schema.data_lock_waits as l join information_schema.tidb_trx as trx on l.current_holding_trx_id = trx.id where l.trx_id = 425497223886536705\G
```

```sql
*************************** 1. row ***************************
               key: 7480000000000000475f728000000000000002
                ID: 425497219115778059
        START_TIME: 2021-06-08 09:04:38.292000
CURRENT_SQL_DIGEST: a4e28cc182bdd18288e2a34180499b9404cd0ba07e3cc34b6b3be7b7c2de7fe9
             STATE: LockWaiting
WAITING_START_TIME: 2021-06-08 09:04:38.336264
   MEM_BUFFER_KEYS: 1
  MEM_BUFFER_BYTES: 19
        SESSION_ID: 97
              USER: root
                DB: test
   ALL_SQL_DIGESTS: [0fdc781f19da1c6078c9de7eadef8a307889c001e05f107847bee4cfc8f3cdf3, a4e28cc182bdd18288e2a34180499b9404cd0
ba07e3cc34b6b3be7b7c2de7fe9, a4e28cc182bdd18288e2a34180499b9404cd0ba07e3cc34b6b3be7b7c2de7fe9]
1 row in set (0.01 sec)
```

> If the `start_ts` of the current transaction is unknown, you can try to find it out from the information in the `TIDB_TRX` / `CLUSTER_TIDB_TRX` table or [`PROCESSLIST` / `CLUSTER_PROCESSLIST`](/information-schema/information-schema-processlist.md) table.

Reviewer 2: ```suggestion
If the `start_ts` of the current transaction is unknown, you can try to find it out from the information in the `TIDB_TRX` / `CLUSTER_TIDB_TRX` table or in the [`PROCESSLIST` / `CLUSTER_PROCESSLIST`](/information-schema/information-schema-processlist.md) table.
```
